File: Python/array/car_fleet.py
from typing import List

def carFleet(target: int, position: List[int], speed: List[int]) -> int:
    if not position or not speed:
        return 0

    remain = {(position[0], (target - position[0]) / speed[0])}
    n = len(position)
    for i in range(1, n):
        remain.add((position[i], (target - position[i]) / speed[i]))
    
    count = n
    remain = sorted(remain)
    flag = remain[-1][1]
    for i in range(1, n):
        if remain[n - 1 - i][1] <= flag:
            count -= 1
        flag = max(flag, remain[n - 1 - i][1])

    return count 


if __name__ == "__main__":
    print(carFleet(target = 12, position = [10,8,0,5,3], speed = [2,4,1,1,3]))

File: core/tmp/Java/src/com/dbc/algorithm/Array/CarFleet.java
package com.dbc.algorithm.Array;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class CarFleet {
    public static int carFleet(int target, int[] position, int[] speed) {
        if (position.length == 0 || speed.length == 0) {
            return 0;
        }

        List<double[]> remain = new ArrayList<>();
        for (int i = 0; i < position.length; i++) {
            remain.add(new double[]{(double) target - position[i], ((double) target - position[i]) / speed[i]});
        }

        Collections.sort(remain, new Comparator<double[]>() {
            @Override
            public int compare(double[] o1, double[] o2) {
                return (int) o1[0] - (int) o2[0];
            }
        });

        double max = remain.get(0)[1];
        int result = position.length;
        for (int i = 1; i < position.length; i++) {
            if (remain.get(i)[1] <= max)
                result--;
            max = Math.max(max, remain.get(i)[1]);
        }

        return result;
    }
}
